Separate fan speeds into percentages and presets modes

bdraco opened the pull request:
Continuation of @Jc2k 's work in https://github.com/home-assistant/core/pull/39204

Architecture issue: https://github.com/home-assistant/architecture/issues/127

WTH: https://community.home-assistant.io/t/wth-homekit-controller-support-for-4-fan-speeds/221305

- [x] Update developer documentation - https://github.com/home-assistant/developers.home-assistant/pull/777
- [x] Update frontend to be a slider that calls `fan.set_percentage` or `fan.set_preset_mode` instead of a speed list. https://github.com/home-assistant/frontend/pull/8216
 
## Breaking change

### For users

The fan entity model has changed to split named speeds into percentages in the range from 0 (off)-100 and preset modes.

#### Why change?

This change allowed us to expand the number of supported speeds to accommodate additional fan models in Home Assistant. (https://github.com/home-assistant/core/pull/45407#issuecomment-765109659)

We had 3 fan speeds and that worked great as long as the fan had no more than 3 speeds

Over time we received a number of requests to add fans with 4, 5, or even more speeds and preset modes.  This put us in the difficult position of having to reject this support because the underlying fan model didn’t support it.

Percentages were chosen because they can represent up to 100 speeds which should accommodate all fans.  Additionally it’s a lot easier to ask your voice voice assistant to set the fan to 20% then remember that medium-low is actually low-medium or have to learn and remember how to say the speeds for every fan model you want to control.

#### Moving forward

Calls in automations and scripts to `fan.set_speed` should be replaced:

- Calls that set a speed should use `fan.set_percentage`.
- Calls that set a preset mode should use `fan.set_preset_mode`.

The following speeds existed in core integration and will now be automatically identified as preset modes: `auto`, `smart`, `interval`, `idle`, and `favorite`

Both of the new calls are backwards compatible.

Calls in automations and scripts to `fan.turn_on` that use the `speed` attribute should be switched to use the `percentage` or `preset_mode` attribute once the underly integration has been updated to support it.

All core fans have been updated to ensure that calls to the `fan.turn_on` service
map `percentage` or `preset_mode` to `speed` for backwards compatibility.

### For developers

The core issues we are solving is:

* non portable interface to speed for voice assistants
* limited about of speeds
* presets and speeds mixed together in speeds

Custom fan integrations should add the `@fan_compat` decorator to
`async_turn_on` or `turn_on` function to ensure backwards compatibility:

Example:
```
from homeassistant.components.fan import fan_compat

...

@fan_compat
async def async_turn_on(self, speed: Optional[str] = None, percentage: Optional[int] = None, preset_mode: Optional[str] = None, **kwargs: Any) -> None:
```

The function signature for `async_turn_on` is now:

`async def async_turn_on(self, speed: str = None, percentage: int = None, preset_mode: Optional[str] = None, **kwargs) -> None:`

The function signature for `turn_on` is now:

`def turn_on(self, speed: str = None, percentage: int = None, preset_mode: Optional[str] = None, **kwargs) -> None:`

Developers that are creating new integrations should review the updated fan entity model documentation at https://developers.home-assistant.io/docs/core/entity/fan

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

At its heart this PR adds support for a new method on fans, `async_set_percentage` (and its sync counterpart).

The goal is to provide a gradual transition and allow the existing interface to continue working by detecting when `async_set_percentage` and `set_percentage` have been overridden and provide a backwards compatible interface.

If a device supports `async_set_percentage` then it doesn't need to implement `async_set_speed` - HA provides a mapping function that converts `medium` into a percentage.

If a device does not support `async_set_percentage` then a mapping function will translate calls to that method into reasonable `async_set_speed` calls.

The `speed` attribute will always be available - using a mapping function if only a percentage is available.

A new `percentage` attribute is added and will also always be available - again using a mapping function where needed - turning `low` into `33%` etc.

The mapping function is dynamic and tries to accommodates fans that don't follow the spec properly (e.g. they have added extra speeds) using a similar mechanism to the one in `homekit`.

The PR ports demo fans to the new API to test it out.



## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16265

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Jc2k committed:
fan: Add an async_set_speed_percentage service

Jc2k committed:
homekit_controller: Port to use new fan API

bdraco committed:
Remove non-fan code

bdraco committed:
Attempt native checks

bdraco committed:
Tweaks

bdraco committed:
add backwards compat to async_turn_on

bdraco committed:
name

bdraco committed:
remove unrelated changes

bdraco committed:
remove unrelated changes

bdraco committed:
cleanup

bdraco committed:
update schema

bdraco committed:
backcompat

bdraco committed:
tweaks

bdraco committed:
Add demo percentage fans

bdraco committed:
update tests

bdraco committed:
tweaks

bdraco committed:
tweaks

bdraco committed:
tweaks

bdraco committed:
tweaks

bdraco committed:
tweaks

bdraco committed:
tweaks

bdraco committed:
tweaks

bdraco committed:
do not base class

bdraco committed:
fix test

bdraco committed:
fix

bdraco committed:
simplify

bdraco committed:
update demo tests

bdraco committed:
Fix handling of invalid speeds

bdraco committed:
remove debug

bdraco committed:
Add comments

bdraco committed:
Add comments

bdraco committed:
Update function signatures for back compat

bdraco committed:
Update function signatures for back compat

bdraco committed:
fix reversed logic

bdraco commented:
Related issues / prs:

https://github.com/esphome/issues/issues/1278
https://github.com/nguyer/homeassistant-bond-home/pull/37
https://github.com/home-assistant/core/pull/37703
https://github.com/home-assistant/core/issues/26180
https://github.com/home-assistant/core/pull/30503
https://github.com/home-assistant/core/pull/29033
https://github.com/home-assistant/core/issues/24606
https://github.com/home-assistant/core/pull/28409
https://github.com/home-assistant/core/pull/26889
https://github.com/home-assistant/architecture/issues/130
https://github.com/home-assistant/core/pull/19767
https://github.com/home-assistant/core/pull/19179#discussion_r243912612
https://github.com/home-assistant/architecture/issues/414
https://github.com/home-assistant/core/pull/36132
https://github.com/home-assistant/core/pull/41412
https://github.com/home-assistant/core/pull/42925
https://github.com/home-assistant/core/issues/42924
https://github.com/Misiu/argon40/issues/9
https://github.com/home-assistant/core/pull/43915
https://github.com/home-assistant/core/issues/45234
https://github.com/mikelawrence/senseme-hacs/issues/13

bdraco committed:
update common tests

Jc2k commented:
I've only had a cursory look through so far but I like what I see. I am a big fan of the gradual transition approach.

What is the breaking change at this stage?

bdraco commented:
> I've only had a cursory look through so far but I like what I see. I am a big fan of the gradual transition approach.
> 
> What is the breaking change at this stage?

Custom integration will have to update their `async_turn_on` to be able to handle `percentage` otherwise it will be ignored when a user calls the `fan.turn_on` service with a `percentage`

Jc2k commented:
Ah ok - that's not too onerous I don't think.

bdraco committed:
Update demo integration order to match docs

bdraco committed:
one more place to update to match the docs

bdraco committed:
fix logic

bdraco committed:
cleanup schema

bdraco commented:
Looks like coverage is needed for backcompat

https://github.com/home-assistant/core/pull/45407/checks?check_run_id=1746793288

bdraco committed:
augment test cover

bdraco committed:
reduce dupe code

bdraco commented:
It might be better to make a '@fan_speed_compat' and '@fan_percentage_compat' decorator instead.  Might need 4 of them though for async vs sync so not sure how much it saves but at least it's more dry

bdraco commented:
Will pick this back up tomorrow, but it's at least ready to get feedback

MartinHjelmare commented on homeassistant/components/fan/__init__.py:
```diff
@@ -182,6 +264,81 @@ def capability_attributes(self):
             return {ATTR_SPEED_LIST: self.speed_list}
         return {}
 
+    def speed_to_percentage(self, speed: str) -> int:
```
```suggestion
    def speed_to_percentage(self, speed: str) -> Optional[int]:
```

MartinHjelmare commented on homeassistant/components/fan/__init__.py:
```diff
@@ -182,6 +264,81 @@ def capability_attributes(self):
+
+        return normalized_speed_list
+
+    def percentage_to_speed(self, value: int) -> str:
```
```suggestion
    def percentage_to_speed(self, value: int) -> Optional[str]:
```

MartinHjelmare commented on homeassistant/components/fan/__init__.py:
```diff
@@ -182,6 +264,81 @@ def capability_attributes(self):
+        return ((speed_offset) * 100) // (speeds_len - 1)
+
+    @property
+    def _normalized_speed_list(self):
```
```suggestion
    def _normalized_speed_list(self) -> List[str]:
```

MartinHjelmare commented on homeassistant/components/fan/__init__.py:
```diff
@@ -105,19 +130,48 @@ async def async_unload_entry(hass, entry):
         if speed == SPEED_OFF:
             await self.async_turn_off()
+        elif not hasattr(self.async_set_percentage, _FAN_NATIVE):
+            await self.async_set_percentage(self.speed_to_percentage(speed))
```
We may need a check here since `speed_to_percentage` may return `None`.

bdraco replied:
I'm not sure if its better to raise a named exception or let it blow up since I don't think there is any way to know what the user wants since *I think* the only way this happens is if the fan has no valid speeds.

MartinHjelmare replied:
I think we should raise already in `speed_to_percentage` and consolidate the return type.

MartinHjelmare commented on homeassistant/components/fan/__init__.py:
```diff
@@ -105,19 +130,48 @@ async def async_unload_entry(hass, entry):
+            await self.hass.async_add_executor_job(self.set_percentage, percentage)
         else:
-            await self.hass.async_add_executor_job(self.set_speed, speed)
+            await self.async_set_speed(self.percentage_to_speed(percentage))
```
Also here `percentage_to_speed` may return `None`.

bdraco replied:
https://github.com/home-assistant/core/pull/45407#discussion_r562832644

bdraco committed:
Update homeassistant/components/fan/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
Update homeassistant/components/fan/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
Update homeassistant/components/fan/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
Update homeassistant/components/fan/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco commented:
I'm hoping to clean this up a bit to use a `@speed_compat` and `@percentage_compat` decorator to make this easier for devs who are transitioning their fans and make this a bit more DRY

Will probably rework the shared comment to something like

# The fan entity model has changed. The @percentage_compat decorator will ensure the speed argument is # set when a percentage is passed in.  When this entity is updated to use the new model and `speed` and # `set_speed` have been removed switch the decorator to @speed_compat to ensure the percentage argument # will be filled for places that still pass in speed instead of percentage.

bdraco committed:
try decorator

bdraco committed:
DRY

bdraco committed:
DRY

bdraco committed:
DRY

bdraco committed:
DRY

bdraco committed:
DRY

bdraco committed:
DRY

bdraco committed:
DRY

bdraco committed:
DRY

bdraco committed:
DRY

bdraco committed:
pylint

bdraco committed:
fix copy/paste error

bdraco committed:
Fix wrapper

bdraco committed:
remove unused code

bdraco commented:
Managed to get this to 100% diff hit 👍

bdraco committed:
raise if there are no valid speeds

MartinHjelmare commented on homeassistant/components/fan/__init__.py:
```diff
@@ -46,17 +48,30 @@
 ATTR_OSCILLATING = "oscillating"
 ATTR_DIRECTION = "direction"
 
+INVALID_SPEEDS_FILTER = {"on", "auto", "smart"}
```
Does this filter cover all the existing fan platforms with invalid speeds?

bdraco replied:
It does for at least the ones I could find and this list https://github.com/home-assistant/architecture/issues/127#issuecomment-678999782

MartinHjelmare commented:
Do we need to limit the mqtt fan platform allowed speed values to make it compatible and have some valid speeds?

bdraco commented:
> Do we need to limit the mqtt fan platform allowed speed values to make it compatible and have some valid speeds?

I think we should remove the ability for `mqtt` to set a non-allowed speed value.  Right now its accepting an arbitrary string to `async_set_speed`

bdraco commented:
It looks like zha has a fan speed on `on`

bdraco commented:
```
# Additional speeds in zigbee's ZCL
# Spec is unclear as to what this value means. On King Of Fans HBUniversal
# receiver, this means Very High.
SPEED_ON = "on"
```
apparently SPEED_ON means very high

bdraco committed:
Adjust _normalized_speed_list to handle existing exceptions

bdraco committed:
Add percentage_compat to zwavejs

bdraco commented:
rebased to pickup zwavejs

bdraco committed:
works better with the right keyword

MartinHjelmare commented:
I think we need to address template fans too and validate the items in the configured speed list.

MartinHjelmare commented on homeassistant/components/fan/__init__.py:
```diff
@@ -46,17 +48,36 @@
+_INVALID_SPEED_ON = "on"
+_INVALID_SPEED_AUTO = "auto"
+_INVALID_SPEED_SMART = "smart"
+_INVALID_SPEEDS_FILTER = {_INVALID_SPEED_ON, _INVALID_SPEED_AUTO, _INVALID_SPEED_SMART}
```
We need to handle all the modes that xiamoi_miio has too.

MartinHjelmare replied:
Maybe we can just filter away all speeds but the four standard speeds instead of the inverse?

bdraco replied:
Yikes

```
OPERATION_MODES_AIRPURIFIER = ["Auto", "Silent", "Favorite", "Idle"]
OPERATION_MODES_AIRPURIFIER_PRO = ["Auto", "Silent", "Favorite"]
OPERATION_MODES_AIRPURIFIER_PRO_V7 = OPERATION_MODES_AIRPURIFIER_PRO
OPERATION_MODES_AIRPURIFIER_2S = ["Auto", "Silent", "Favorite"]
OPERATION_MODES_AIRPURIFIER_3 = ["Auto", "Silent", "Favorite", "Fan"]
OPERATION_MODES_AIRPURIFIER_V3 = [
    "Auto",
    "Silent",
    "Favorite",
    "Idle",
    "Medium",
    "High",
    "Strong",
]
OPERATION_MODES_AIRFRESH = ["Auto", "Silent", "Interval", "Low", "Middle", "Strong"]
```

If we filter `Auto`, `Interval`, `Idle`, `Favorite`, `Idle` we should get an ordered list from lowest to highest for all of them

MartinHjelmare replied:
Is there a reason we don't just filter on the valid speeds instead of the invalid?

bdraco replied:
Yes, the goal is to get the speeds in a list from lowest to highest by removing speeds that are not valid or out of order so we can map them to percentages.

Since there are so many integration (and custom ones) that have worked around the entity model it seemed like we should try to keep them working though the transition period.

MartinHjelmare replied:
Right, but wouldn't we achieve that by filtering in the valid speeds just the same as filtering out the invalid speeds?

bdraco replied:
I did some renaming and added examples to the code.  Hopefully I did a better job communicating what I'm trying to achieve here with the last commit

bdraco commented:
> I think we need to address template fans too and validate the items in the configured speed list.

It looks like we have explicit tests to make sure custom speeds work. 🤔 

https://github.com/home-assistant/core/blob/dev/tests/components/template/test_fan.py#L512
https://github.com/home-assistant/core/blob/dev/tests/components/template/test_fan.py#L686

bdraco committed:
Augment invalid speed list

bdraco commented:
Maybe we need a separate utility for converting an speed list to a percentage and back to help integration authors with conversions for when the underlying library has a list of speeds. Ones like isy and zwave will be easy but ones with 4,5,6.. etc speeds will likely need some help.

bdraco committed:
breakout percentage code

bdraco commented:
I've added the utility but still need to add unit tests for it.

Once I do that I'll update the devs docs for tips on how to convert existing fans that have ranged value and speed lists

bdraco committed:
fix paste error

bdraco committed:
Coverage

bdraco committed:
add hints

bdraco commented:
I've updated the developer documentation to provide guidance on the utility to make it easier for devs to convert.

bdraco committed:
naming

bdraco committed:
reorder for readability

bdraco commented:
I started the template fan changes here: https://github.com/home-assistant/core/pull/45478

bdraco committed:
Cover

bdraco commented:
Should be ready for another pass.  Back to 💯 coverage

MartinHjelmare commented:
The main concern I have is how we're breaking the integrations for users. With the gradual transition approach we're breaking integrations by removing support for some speeds and this is unpredictable for users. It's not clear to me how we'll communicate this and if this is the best approach.

bdraco commented:
> The main concern I have is how we're breaking the integrations for users. With the gradual transition approach we're breaking integrations by removing support for some speeds and this is unpredictable for users. It's not clear to me how we'll communicate this and if this is the best approach.

The user can still access the non-speed speeds via a service call since they are only filtered for the purposes of converting to the percentage.

I could update the frontend PR to so they have both the percentage slider and the drop down. This would allow them to continue to access the non-speed speeds.

In the future we could introduce the concept of presets similar to climate to accommodate: auto/smart/interval/favorite. Then we remove the drop down from the UI and replace it with the preset one.

bdraco commented:
The more I think about this, the core issues we are solving is:

* non portable interface to speed for voice assistants
* limited about of speeds
* presets and speeds mixed together in speeds

As much as I'd prefer to do this in steps, the tail is long and it's probably not reasonable to keep a by the end of 2021 timeline if we expect integration authors to have to update a second time to add presets.

It probably makes sense to separate the presets now and take the pain one time instead of twice.

MartinHjelmare commented:
Yes, that's a good point. I suggest we take some more views from other members before committing in any direction.

bdraco commented:
> Yes, that's a good point. I suggest we take some more views from other members before committing in any direction.

Posted here https://github.com/home-assistant/architecture/issues/127#issuecomment-766354367 and cced all members who previously participated.

bdraco committed:
Preset modes for fans (first pass)

Preset modes for fans (first pass) - 2

Preset modes for fans (first pass) - 3

Preset modes for fans (first pass) - 4

Preset modes for fans (first pass) - 5

Preset modes for fans (first pass) - 6

Preset modes for fans (first pass) - 7

Preset modes for fans (first pass) - 8

Preset modes for fans (first pass) - 9

Preset modes for fans (first pass) - 10

Preset modes for fans (first pass) - 11

Preset modes for fans (first pass) - 12

Preset modes for fans (first pass) - 13

Preset modes for fans (first pass) - 14

Preset modes for fans (first pass) - 15

Preset modes for fans (first pass) - 16

Preset modes for fans (first pass) - 17

Preset modes for fans (first pass) - 18

Preset modes for fans (first pass) - 19

Preset modes for fans (first pass) - 20

Preset modes for fans (first pass) - 21

Preset modes for fans (first pass) - 22

Preset modes for fans (first pass) - 23

Preset modes for fans (first pass) - 24

Preset modes for fans (first pass) - 25

Preset modes for fans (first pass) - 26

Preset modes for fans (first pass) - 27

Preset modes for fans (first pass) - 28

Preset modes for fans (first pass) - 29

Preset modes for fans (first pass) - 30

Preset modes for fans (first pass) - 31

Preset modes for fans (first pass) - 32

Preset modes for fans (first pass) - 33

Preset modes for fans (first pass) - 34

Improve exception error messages

coverage

Fix backwards compat

DRY

Update test

Tests for valid presets

bdraco commented:
Giving preset_mode a shot to make sure its doable. If there is some feedback about not going in this direction in the arch issue, I'll revert and go another way.

jbouwh commented:
Are you also updating mqtt fan with autodiscovery?

bdraco commented:
> Are you also updating mqtt fan with autodiscovery? 

This is an entity model change with transition compatibility only. Each integration will need to update in the manner that best works for it.

bdraco commented:
Back to 100% diff hit.

bdraco commented:
I've updated the frontend PR as well

bdraco commented:
Updated the template PR https://github.com/home-assistant/core/pull/45478 (could use a bit more coverage, but manually testing passed)

bdraco commented:
Updated the developer docs https://github.com/home-assistant/developers.home-assistant/pull/777

bdraco commented:
Update docs https://github.com/home-assistant/home-assistant.io/pull/16265

bdraco commented:
Template fan and template fan docs updated https://github.com/home-assistant/core/pull/45478 with 100% diff hit

bdraco commented:
I think I can get rid of the decorator and not change any integration by pointing the service registration at a wrapper for turn_on and set_speed

bdraco committed:
Git rid of decorator -- attempt 1

bdraco commented:
Looks like wilight added a fan.

Will rebase again

bdraco committed:
update wilight

bdraco committed:
Template was not conforming

bdraco commented:
I wanted to see how it would be to follow the dev instructions so I updated `bond` using them.
https://github.com/home-assistant/core/pull/45534

I'm going to make some tweaks now

bdraco committed:
Handle the case where they only implement percentage and not preset_mode

bdraco commented:
I'm going to do one more to make sure the instructions are easy to follow

bdraco commented:
Did isy994 here https://github.com/home-assistant/core/pull/45536

These actually go quick since I know these integrations

bdraco committed:
Increase cover

bdraco committed:
Tweak percentage conversion to make range_value function like ordered_list

bdraco commented:
isy994 and bond resulted in some tweaks.  I'm going to do one that uses an ordered list of speeds to make sure there are no more tweaks.

bdraco commented:
lutron_caseta was easier since it used a named list
https://github.com/home-assistant/core/pull/45540

No tweaks

bdraco commented:
At this point I'm happy with this, and I think its ready for another review pass.

bdraco committed:
Fix speed lookup

bdraco commented:
I'm doing zwave and zwavejs now to make sure there aren't any other adjustments needed

Jc2k commented on homeassistant/components/fan/__init__.py:
```diff
@@ -196,10 +504,81 @@ def state_attributes(self) -> dict:
         if supported_features & SUPPORT_SET_SPEED:
             data[ATTR_SPEED] = self.speed
+            data[ATTR_PERCENTAGE] = self.percentage
+            data[ATTR_PRESET_MODE] = self.preset_mode
```
Can a fan have presets but not speed controls? I vaguely recall there was at least one fan like that 😤 in which case should this be separated out under its own supported feature?

bdraco replied:
I could add a new flag and have it only take effect once they have implemented `set_percentage` or `set_preset_mode` in their integration, otherwise it will look at `SUPPORT_SET_SPEED` for backwards compatibility.

bdraco replied:
That's going to be an odd behavior change though when they do the conversion.

For maximum compatibility, we could check if there are `preset_modes` and only show the attribute if there are.

bdraco replied:
I started going down that rabbit hole and I think this would need to come later in another update to the model as it would break backwards compatibility since the integrations would have to add the new `SUPPORT_SET_PRESET_MODE`.

bdraco replied:
We can hide the attribute if there are no entries in `self.preset_modes`. I don't think it will break anything functional (just some tests to update).

Jc2k replied:
I was more thinking about the other direction:

https://github.com/home-assistant/core/blob/dev/homeassistant/components/xiaomi_miio/fan.py#L797
https://github.com/home-assistant/core/blob/dev/homeassistant/components/xiaomi_miio/fan.py#L369

A Xiaomi Miio Air Purifier Pro doesnt actually have *any* speeds, *everything* will be in `self.preset_modes`. What happens for one of those? We still have to support "Silent" as a speed for compatibility, I guess. But are there any other gotchas.

bdraco replied:
I'll add a demo fan that has only presets and no speeds to make sure everything is handled in a sane way

bdraco replied:
To support this I'm going to have to add `SUPPORT_PRESET_MODES`.

bdraco replied:
I've added this and a demo fan that only does preset modes.

I may need to tweak a few tests once the ci runs

bdraco replied:
<img width="408" alt="Screen Shot 2021-01-25 at 3 09 33 PM" src="https://user-images.githubusercontent.com/663432/105766479-69e25e80-5f1f-11eb-9855-a0ce7f753749.png">

bdraco replied:
I updated the developer docs pr for this as well

bdraco replied:
Tested on one of my instances and looks good

bdraco committed:
Add SUPPORT_PRESET_MODE

bdraco committed:
Update tests for preset only fan

bdraco commented:
Testing on production instances

- [ ] a8128
- [ ] a8123
- [ ] m8123
- [ ] m8125
- [ ] m8127

bdraco commented:
I pulled out Jc2k's original work to update homekit_controller from #39204 and updated it in https://github.com/home-assistant/core/pull/45547 to verify I didn't break anything in the 1:1 case.

bdraco commented:
I've done all the manual testing I can think to do with this and ran though the code line by line again.

bdraco commented:
Did another run though. Can't find anything else that needs to be adjusted.

bdraco commented:
I've converted 10 out of the 27 fans in core in separate PRs.

I was able to make all 10 of them conform to the spec without having to the ability to set some speeds.

Jc2k commented:
This is looking really good. My +1 doesn't mean much outside of homekit_controller, but you have it. I can't think of a better way we could handle the transition AND cope with the non compliant fans.

bdraco commented:
I've got HomeKit updated in 

Alexa and Google both support the percentage model and they looks quite strait-forward to update 

Alexa percentage controller

Google supportsFanSpeedPercent

https://developers.google.com/assistant/smarthome/traits/fanspeed

MartinHjelmare commented:
I'll look at this again after the beta cut.

balloob reviewed (approved):
Changes look good!

balloob commented:
This should not be merged until we cut the beta.

bdraco committed:
Update homeassistant/components/demo/fan.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
Update homeassistant/components/demo/fan.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco commented on homeassistant/components/demo/fan.py:
```diff
@@ -100,17 +187,124 @@ def oscillate(self, oscillating: bool) -> None:
+        if preset_mode in self.preset_modes:
+            self._preset_mode = preset_mode
+            self._percentage = None
+            self.async_write_ha_state()
```
Thanks for catching that.

bdraco commented:
Final testing looks good 👍
